devito/function.py

```diff
@@ -1074,3 +1122,107 @@ def __indices_setup__(cls, **kwargs):
+
+
+class PrecomputedSparseFunction(AbstractSparseFunction):
+    is_PrecomputedSparseFunction = True
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            super(PrecomputedSparseFunction, self).__init__(*args, **kwargs)
+
+            # Grid points per sparse point (2 in the case of bilinear and trilinear)
+            r = kwargs.get('r')
+            if not isinstance(r, int) and r > 0:
+                raise ValueError('Interpolation requires parameter `r` (>0)')
+            self.r = r
+
+            gridpoints = Function(name="%s_gridpoints" % self.name,
+                                  dimensions=(self.indices[-1], Dimension(name='d')),
+                                  shape=(self.npoint, self.grid.dim), space_order=0,
+                                  dtype=np.int32)
+
+            gridpoints_data = kwargs.get('gridpoints', None)
+            assert(gridpoints_data is not None)
+            gridpoints.data[:] = gridpoints_data[:]
+            self.gridpoints = gridpoints
+
+            coefficients = Function(name="%s_coefficients" % self.name,
+                                    dimensions=(self.indices[-1], Dimension(name='d'),
+                                                Dimension(name='i')),
+                                    shape=(self.npoint, self.grid.dim, self.r),
+                                    space_order=0)
+            coefficients_data = kwargs.get('coefficients', None)
+            assert(coefficients_data is not None)
+            coefficients.data[:] = coefficients_data[:]
+            self.coefficients = coefficients
+
+    def interpolate(self, expr, offset=0, u_t=None, p_t=None, cummulative=False):
+        """Creates a :class:`sympy.Eq` equation for the interpolation
+        of an expression onto this sparse point collection.
+
+        :param expr: The expression to interpolate.
+        :param offset: Additional offset from the boundary for
+                       absorbing boundary conditions.
+        :param u_t: (Optional) time index to use for indexing into
+                    field data in `expr`.
+        :param p_t: (Optional) time index to use for indexing into
+                    the sparse point data.
+        :param cummulative: (Optional) If True, perform an increment rather
+                            than an assignment. Defaults to False.
+        """
+        expr = indexify(expr)
+
+        # Apply optional time symbol substitutions to expr
+        if u_t is not None:
+            time = self.grid.time_dim
+            t = self.grid.stepping_dim
+            expr = expr.subs(t, u_t).subs(time, u_t)
+
+        coefficients = self.coefficients.indexed
+        gridpoints = self.gridpoints.indexed
+        p, _, _ = self.coefficients.indices
+        dim_subs = []
+        coeffs = []
+        for i, d in enumerate(self.grid.dimensions):
+            rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
```

Comment:
Dow we really need a DInmension for that. Unless go to like 100 points in each direction, can just write down the r equations.

Reply:
`Dimension`s are the only way to structure loops in devito. With a loop, the generated code is independent of the interpolator used. 
This is a WIP anyway. I don't think one way is better than the other. The best would be to pass the size information to the compiler and let the compiler decide whether it wants to unroll the loop.